src/prefect/orion/services/flow_run_notifications.py

```diff
@@ -34,9 +31,14 @@ async def run_once(self, db: OrionDBInterface):
         async with session:
             async with session.begin():
                 while True:
+                    # Drain the queue one entry at a time, because if a database error
+                    # happens while sending a notification, we'll need to rollback the
+                    # transaction, which effectively re-queues any notifications that
+                    # we pulled here.  If we drain in batches larger than 1, we risk
+                    # double-sending notifications when a transient error occurs.
                     notifications = await db.get_flow_run_notifications_from_queue(
                         session=session,
-                        limit=self.batch_size,
+                        limit=1,
```

Comment:
are we worried about 
- performance of doing this one by one?
- clearing "bad" notifications from the queue? does this actually delete it? 🤔 

alternatively we could
- open a transaction
- get notifications from queue (which removes them)
- commit
- open a new transaction for looking up the block document and sending the notification etc

Reply:
> are we worried about
> * performance of doing this one by one?

I am worried about that.  I'd like to add some monitoring to the flow_run_notifications queue depth to make sure we are able to keep it drained.

> * clearing "bad" notifications from the queue? does this actually delete it? 🤔

I don't think I'm worried about that one.  This kind of comes back to the subtlety of the error I'm handling here.  Any "normal" error we see here (like a problem with a bad Slack token, for example) will _not_ cause this transaction to roll back, so we won't re-enqueue those.  The kinds of errors I'm catching here are ones that _invalidate the connection_, like transient DB and environmental errors.  In those cases, the transaction is actually _already rolled back_, so we just need to get the Python session objects synced back up with that reality.

I think my comments in the code are a little inaccurate, I'll tweak them.

Reply:
> alternatively we could
> * open a transaction
> * get notifications from queue (which removes them)
> * commit
> * open a new transaction for looking up the block document and sending the notification etc

Jake suggested this in my other version of the PR, to use a separate session and transaction for the notification part.  The only thing I was worried about there was that a transient error (connection drop, db restart, etc) would end up with us not sending a user notification when we should have.  I guess we're running into the whole "queue guarantees" problem here.  I'm trying to make sure we deliver at least once, rather than at most once.